### PR TITLE
Put std::convert::Infallible inside stub::MmapInner

### PR DESCRIPTION
**Pull request by coolreader18**

### CODE REVIEW

**Comment by RazrFalcon** [comment]
Can you elaborate?

File: src/stub.rs [hunk 1]
```diff
@@ -1,10 +1,9 @@
+use std::convert::Infallible as Never;
```
Comment (adamreichold) [inline src/stub.rs:1]: Does renaming `Infallible` gain anything compared to defining `enum Never {}` here?

**Comment by adamreichold** [comment]
> Can you elaborate?

I think this is about making the "self unconstructable" part statically visible to the compiler so that no-code is generated for the unreachable methods.

File: src/stub.rs (hunk 1 above)
Comment (coolreader18) [reply src/stub.rs:1]: No, other than that `Infallible` will "eventually" become a type alias for `!`, once that stabilizes.

Comment (coolreader18) [reply src/stub.rs:1]: If you'd like me to switch to a `enum Never {}` I'm good with that too.

**Comment by coolreader18** [comment]
Yep, that's exactly it. I noticed the comment about `!` and realized it could be stabilized today, it doesn't have to wait til never.

**Comment by coolreader18** [comment]
(something something they call it the never type cause that's when it'll be stabilized)

File: src/stub.rs (hunk 1 above)
Comment (adamreichold) [reply src/stub.rs:1]: While having an alias called "infallible" for the never type when applied to errors, I think it is a bit confusing to use that error-oriented alias in this situation where no errors are involved and we just want to make our type uninhabited.

Personally, I think defining an uninhabited enum here would be clearer.

Comment (coolreader18) [reply src/stub.rs]: Alright, done.

**Review by adamreichold — APPROVED** [review]

**Comment by RazrFalcon** [comment]
Can you also add a comment explaining this trick? Explaining what it is, why it's needed and what are the benefits of it. Or maybe there is a documentation somewhere we can point to.

**Comment by adamreichold** [comment]
https://doc.rust-lang.org/stable/std/primitive.never.html has some information, but I think it does cover this use case exactly.

**Comment by coolreader18** [comment]
I could link to [the docs for the `never` crate](https://docs.rs/never/latest/never/enum.Never.html), which is basically just this.

**Comment by RazrFalcon** [comment]
Yes, this is a good explanation. But why it's better than a panic?

**Comment by adamreichold** [comment]
> Yes, this is a good explanation. But why it's better than a panic?

A panic is code that unwinds or aborts the process, but with this approach there is actually no code generated for these methods if they are called from dependent code.

So to speak, using `Never`, we can establish via the type system that `unreachable_unchecked` is safe to call which is what the empty match boils down to.

**Pull request merged**